https://www.kaggle.com/c/bike-sharing-demand/data?select=train.csv

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

In [2]:
df = pd.read_csv('data1/bike-sharing-demand/train.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   casual      10886 non-null  int64  
 10  registered  10886 non-null  int64  
 11  count       10886 non-null  int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 1020.7+ KB


In [125]:
dropped_df = df.drop(['datetime','casual','registered'], axis=1)
dropped_df

,season,holiday,workingday,weather,temp,atemp,humidity,windspeed,count
0,1,0,0,1,9.84,14.395,81,0.0000,16
1,1,0,0,1,9.02,13.635,80,0.0000,40
2,1,0,0,1,9.02,13.635,80,0.0000,32
3,1,0,0,1,9.84,14.395,75,0.0000,13
4,1,0,0,1,9.84,14.395,75,0.0000,1
...,...,...,...,...,...,...,...,...,...
10881,4,0,1,1,15.58,19.695,50,26.0027,336
10882,4,0,1,1,14.76,17.425,57,15.0013,241
10883,4,0,1,1,13.94,15.910,61,15.0013,168
10884,4,0,1,1,13.94,17.425,61,6.0032,129


In [126]:
X = dropped_df.drop('count',axis=1).values
y = dropped_df['count']

In [ ]:
from sklearn.preprocessing import StandardScaler

In [127]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [128]:
lr = LinearRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)

In [129]:
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

rmse, mse

(np.float64(155.13350586282306), np.float64(24066.40464129056))

In [130]:
r2_score(y_test, y_pred)

np.float64(0.261308860174165)

In [131]:
pd.Series(data = np.abs(np.round(lr.coef_,1)), index=dropped_df.drop('count',axis=1).columns).sort_values(ascending=False)

season        23.3
atemp          5.5
weather        5.4
holiday        4.7
humidity       3.0
temp           1.9
windspeed      0.9
workingday     0.8
dtype: float64

In [114]:
X = dropped_df.drop(['count', 'workingday', 'windspeed', 'temp'],axis=1).values
y = dropped_df['count']

In [115]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [94]:
lr = LinearRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)

In [95]:
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

rmse, mse

(np.float64(155.0761352604271), np.float64(24048.607727310282))

In [67]:
r2_score(y_test, y_pred)

np.float64(0.26185511637110925)

In [68]:
from sklearn.model_selection import cross_val_score
neg_mse_scores = cross_val_score(lr, X, y, scoring='neg_mean_squared_error', cv=5)
neg_mse_scores

array([-13698.21233362, -24507.81660116, -16066.89762602, -40581.7006381 ,
       -45481.08991632])

In [69]:
RMSE = np.sqrt(neg_mse_scores * (-1))
np.mean(RMSE), RMSE

(np.float64(163.0112633521338),
 array([117.03936233, 156.54972565, 126.75526666, 201.44900257,
        213.26295955]))

In [70]:
r2_scores = cross_val_score(lr, X, y, scoring='r2', cv=5)
r2_scores, np.mean(r2_scores)

(array([-0.38223363, -0.08848212,  0.03326709,  0.12877392,  0.03933621]),
 np.float64(-0.05386770591695358))

In [96]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(random_state=42, max_depth=8)
rf.fit(X_train,y_train)
y_pred = rf.predict(X_test)

In [97]:
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

rmse, mse

(np.float64(147.79352714154993), np.float64(21842.926664940052))

In [73]:
r2_score(y_test, y_pred)

np.float64(0.3295560082300917)

In [74]:
neg_mse_scores = cross_val_score(rf, X, y, scoring='neg_mean_squared_error', cv=5)
neg_mse_scores

array([-13541.2665182 , -25928.63673041, -18667.1841574 , -43195.5744361 ,
       -43499.85283743])

In [75]:
RMSE = np.sqrt(neg_mse_scores * (-1))
np.mean(RMSE), RMSE

(np.float64(166.08404005369826),
 array([116.36694771, 161.02371481, 136.62790402, 207.83545038,
        208.56618335]))

In [76]:
r2_scores = cross_val_score(rf, X, y, scoring='r2', cv=5)
r2_scores, np.mean(r2_scores)

(array([-0.36639683, -0.15158596, -0.12319016,  0.07265811,  0.08118443]),
 np.float64(-0.09746608189335666))

현재까지 랜덤포레스트 단독 모델이 가장 낫다

| 모델         | 평가 지표 | 단독 예측 (Individual Prediction) | 교차 검증 (Cross-Validation) |
| :----------- | :-------- | :-------------------------------- | :--------------------------- |
| **리니어 모델** | RMSE      | 155.08                            | 163.01                       |
|              | R2        | 0.2619                            | -0.0539                      |
| **랜덤 포레스트** | RMSE      | 147.79                            | 166.08                       |
|              | R2        | 0.3296                            | -0.0975                      |

In [45]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures

In [116]:
results = []
for degree in range(1,7):
    model_poly = Pipeline([
        ('poly',PolynomialFeatures(degree=degree, include_bias=False)),
        ('linear',LinearRegression())
    ])
    model_poly.fit(X_train, y_train)
    pred_poly = model_poly.predict(X_test)
    mse = mean_squared_error(y_test,pred_poly)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, pred_poly)
    results.append({
        'degree' : degree,
        'MSE' : mse,
        'RMSE' : rmse,
        'R2' : r2
    })
pd.DataFrame(results)

,degree,MSE,RMSE,R2
0,1,24048.607727,155.076135,0.261855
1,2,23347.372935,152.798472,0.283379
2,3,22876.278692,151.249062,0.297838
3,4,22358.819862,149.528659,0.313721
4,5,22387.713012,149.625242,0.312834
5,6,138562.448434,372.239773,-3.253018


다항회귀모델(선형)에서는 4차가 가장 높다

In [47]:
results = []
for degree in range(1,6):
    model_poly = Pipeline([
        ('poly',PolynomialFeatures(degree=degree, include_bias=False)),
        ('RF',RandomForestRegressor(random_state=22, max_depth=8))
    ])
    model_poly.fit(X_train, y_train)
    pred_poly = model_poly.predict(X_test)
    mse = mean_squared_error(y_test,pred_poly)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, pred_poly)
    results.append({
        'degree' : degree,
        'MSE' : mse,
        'RMSE' : rmse,
        'R2' : r2
    })
pd.DataFrame(results)

,degree,MSE,RMSE,R2
0,1,21891.336106,147.957210,0.328070
1,2,21893.763759,147.965414,0.327996
2,3,21961.988551,148.195778,0.325902
3,4,21991.710673,148.296024,0.324989
4,5,21936.983246,148.111388,0.326669


다항회귀모델(랜덤포레스트)에서는 단항이 가장 높다

결론적으로 랜덤포레스트 단독 모델일때 성능이 가장 높았다.

In [134]:
from sklearn.ensemble import GradientBoostingRegressor

In [135]:
gb_clf = GradientBoostingRegressor()
gb_clf.fit(X_train,y_train)
gb_pred = gb_clf.predict(X_test)

In [136]:
mse = mean_squared_error(y_test,gb_pred)
rmse = np.sqrt(mse)

rmse, mse

(np.float64(148.65826046661388), np.float64(22099.27840495962))

In [137]:
r2_score(y_test, gb_pred)

np.float64(0.32168758077472925)

In [138]:
from xgboost import XGBRegressor

In [139]:
xgb = XGBRegressor(n_estimators=400, learning_rate=0.1, max_depth=3, use_label_encoder=False)
evals = [(X_test_scaled, y_test)]
xgb.fit(X_train_scaled, y_train, early_stopping_rounds=40, 
        eval_set=evals, verbose=True)
xgb_pred = xgb.predict(X_test_scaled)

[0]	validation_0-rmse:242.81698
[1]	validation_0-rmse:228.58793
[2]	validation_0-rmse:216.44106
[3]	validation_0-rmse:205.91504
[4]	validation_0-rmse:197.00860
[5]	validation_0-rmse:189.31892
[6]	validation_0-rmse:182.73839
[7]	validation_0-rmse:177.20154
[8]	validation_0-rmse:172.73875
[9]	validation_0-rmse:168.92128
[10]	validation_0-rmse:165.81172
[11]	validation_0-rmse:163.09597
[12]	validation_0-rmse:160.92956
[13]	validation_0-rmse:159.03686
[14]	validation_0-rmse:157.58634
[15]	validation_0-rmse:156.24149
[16]	validation_0-rmse:155.12871
[17]	validation_0-rmse:154.28972
[18]	validation_0-rmse:153.50600
[19]	validation_0-rmse:152.87736
[20]	validation_0-rmse:152.38217
[21]	validation_0-rmse:151.95435
[22]	validation_0-rmse:151.62822
[23]	validation_0-rmse:151.23281
[24]	validation_0-rmse:150.91682
[25]	validation_0-rmse:150.73322
[26]	validation_0-rmse:150.60042
[27]	validation_0-rmse:150.48316
[28]	validation_0-rmse:150.29558
[29]	validation_0-rmse:150.20399
[30]	validation_0-rm

c:\Users\Admin\miniconda3\envs\hi_ml_env\lib\site-packages\xgboost\sklearn.py:793: UserWarning: `early_stopping_rounds` in `fit` method is deprecated for better compatibility with scikit-learn, use `early_stopping_rounds` in constructor or`set_params` instead.
  warnings.warn(


[161]	validation_0-rmse:148.32662
[162]	validation_0-rmse:148.33808
[163]	validation_0-rmse:148.33535
[164]	validation_0-rmse:148.35193
[165]	validation_0-rmse:148.34697
[166]	validation_0-rmse:148.34128
[167]	validation_0-rmse:148.34844
[168]	validation_0-rmse:148.34425
[169]	validation_0-rmse:148.35659
[170]	validation_0-rmse:148.36644
[171]	validation_0-rmse:148.36879
[172]	validation_0-rmse:148.37582
[173]	validation_0-rmse:148.37943
[174]	validation_0-rmse:148.38102
[175]	validation_0-rmse:148.37711
[176]	validation_0-rmse:148.37004
[177]	validation_0-rmse:148.36615


In [140]:
mse = mean_squared_error(y_test,xgb_pred)
rmse = np.sqrt(mse)

rmse, mse

(np.float64(148.22570905944775), np.float64(21970.860826176053))

In [141]:
r2_score(y_test, xgb_pred)

np.float64(0.32562921348053675)